src/Compiler/Mapper/Object/MapDiscriminatedObject.php

```diff
@@ -0,0 +1,160 @@
+<?php declare(strict_types = 1);
+
+namespace ShipMonk\InputMapper\Compiler\Mapper\Object;
+
+use Attribute;
+use LogicException;
+use Nette\Utils\Arrays;
+use PhpParser\Node\Expr;
+use PHPStan\PhpDocParser\Ast\Type\GenericTypeNode;
+use PHPStan\PhpDocParser\Ast\Type\IdentifierTypeNode;
+use PHPStan\PhpDocParser\Ast\Type\TypeNode;
+use ShipMonk\InputMapper\Compiler\CompiledExpr;
+use ShipMonk\InputMapper\Compiler\Mapper\GenericMapperCompiler;
+use ShipMonk\InputMapper\Compiler\Mapper\Scalar\MapString;
+use ShipMonk\InputMapper\Compiler\Php\PhpCodeBuilder;
+use ShipMonk\InputMapper\Compiler\Type\GenericTypeParameter;
+use ShipMonk\InputMapper\Runtime\Exception\MappingFailedException;
+use function array_keys;
+use function count;
+use function ucfirst;
+
+/**
+ * @template T of object
+ */
+#[Attribute(Attribute::TARGET_PARAMETER | Attribute::TARGET_PROPERTY)]
+class MapDiscriminatedObject implements GenericMapperCompiler
+{
+
+    /**
+     * @param class-string<T> $className
+     * @param array<string, class-string> $subtypeMapping
+     * @param list<GenericTypeParameter> $genericParameters
+     */
+    public function __construct(
+        public readonly string $className,
+        public readonly string $discriminatorFieldName,
+        public readonly array $subtypeMapping,
+        public readonly array $genericParameters = [],
+    )
+    {
+    }
+
+    public function compile(Expr $value, Expr $path, PhpCodeBuilder $builder): CompiledExpr
+    {
+        $provider = $builder->propertyFetch($builder->var('this'), 'provider');
```

Comment:
You should add safety check that all the subtypes are in fact subtypes of `$this->className`

* `\ShipMonk\InputMapper\Compiler\Type\PhpDocTypeUtils::isSubTypeOf`
* `\ShipMonk\InputMapper\Compiler\Exception\CannotCompileMapperException`

Reply:
Added